### PR TITLE
font shorthand should reset more longhands

### PR DESCRIPTION
**Pull request by darinadler**
#### 9de002d111c74dfffa39582a67d6af89b9abf21c
<pre>
font shorthand should reset more longhands
<a href="https://bugs.webkit.org/show_bug.cgi?id=246912">https://bugs.webkit.org/show_bug.cgi?id=246912</a>
rdar://101486558

Reviewed by Tim Nguyen.

<a href="https://drafts.csswg.org/css-fonts/#font-prop">https://drafts.csswg.org/css-fonts/#font-prop</a>
&quot;All subproperties of the font property are first reset to their initial values&quot;

There are multiple things done here:
1) Add more longhands to font shorthand for all &quot;subproperties of font property&quot; listed.
2) Change style building code to reset them to initial values.
3) Change computed style code to make an empty font shorthand when any of these have non-initial values.
4) Use CSSPrimitiveValue for font-style keywords without angles, rather than CSSFontStyleValue.
5) Renamed CSSFontStyleValue to CSSFontStyleWithAngleValue.

The change (4) simplifies some code. The legacy CSS object model doesn&apos;t otherwise represent a
simple keyword as a CSSValue subclass. In an earlier version of this work, the rest of the
changes depended on this, but at this point it&apos;s just a separate improvement.

* LayoutTests/fast/css/font-shorthand-expected.txt: Updated since many more properties are set
to initial by the font shorthand now.

* LayoutTests/fast/css/font-systemFontID-parsing-expected.txt: Updated since the test is changed,
but also to expect a wrong result due to a bug we haven&apos;t fixed: &quot;font: caption&quot; does not serialize.
* LayoutTests/fast/css/font-systemFontID-parsing.html: Updated to match what&apos;s specified, where
font-family and other font subproperties do not reflect system fonts, which can only be specified
with the font shorthand (which makes it not quite a shorthand).

* LayoutTests/fast/inspector-support/style-expected.txt: Updated since many more properties are set
to initial by the font shorthand now.

* LayoutTests/fast/text/font-stretch-parse-expected.txt: Updated.
* LayoutTests/fast/text/font-stretch-parse.html: Updated to expect font shorthand to serialize
as empty string when font-stretch is set to a value font shorthand can&apos;t express.

* LayoutTests/fast/text/font-style-parse-expected.txt: Updated.
* LayoutTests/fast/text/font-style-parse.html: Updated to expect font shorthand to serialize
as empty string when font-stretch is set to a value font shorthand can&apos;t express. Also took
out &quot;window.&quot; so test is slightly less wordy.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-palette-vs-shorthand-expected.txt:
Expect all PASS instead of FAIL.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/system-fonts-serialization.tentative-expected.txt:
Expect a different failure; this test expects all subproperties of font to serialize as the
empty string when the font property is set to a system font. This change doesn&apos;t do that for
all of them, so the test doesn&apos;t pass yet, but it fails later.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
Updated since the font shorthand now affects so many more font subproperties. The test is still
failing because we haven&apos;t added the font shorthand to the StyleProperties::asTextInternal function,
and this will need to remain until we revise editing code to not depend on current behavior.

* LayoutTests/platform/win/fast/css/font-shorthand-expected.txt: Added. This is different because
PLATFORM(WIN) doesn&apos;t yet support variation fonts.
* LayoutTests/platform/win/fast/inspector-support/style-expected.txt: Added. Ditto.

* LayoutTests/svg/css/getComputedStyle-basic-expected.txt: Updated since font-style is now a
CSSPrimitiveValue instead of a CSSFontStyleValue, web-exposed in the legacy CSS object model.

* Source/WebCore/Sources.txt: Renamed to CSSFontStyleWithAngleValue.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Ditto.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::calculateItalicRange): Instead of explicitly checking for CSSFontStyleValue, just call
BuilderConverter::convertFontStyleFromValue on anything other than a CSSFontStyleRangeValue. This
makes this function work with CSSPrimitiveValue, and it&apos;s nicer for encapsulation since this
already needs to know about the CSSFontStyleRangeValue class.

* Source/WebCore/css/CSSFontFaceSet.cpp: Removed include of CSSFontStyleValue.h.

* Source/WebCore/css/CSSFontStyleWithAngleValue.cpp: Renamed from Source/WebCore/css/CSSFontStyleValue.cpp.
(WebCore::CSSFontStyleValue::CSSFontStyleValue): Moved out of header.
Also only take a single primitive value for oblique angle.
(WebCore::CSSFontStyleValue::create): Ditto.
(WebCore::CSSFontStyleValue::customCSSText const): Updated since this class is now only used for
oblique with an angle. Also use makeString instead of StringBuilder.
(WebCore::CSSFontStyleValue::equals const): Updated since class is now only used for oblique
with an angle.
* Source/WebCore/css/CSSFontStyleWithAngleValue.h: Renamed from Source/WebCore/css/CSSFontStyleValue.h.
Changed this to a class that just stores oblique with a non-optional angle since we now use
CSSPrimitiveValue directly for keywords without angles. Also removed isItalicOrOblique function,
since this is an editing heuristic that should have been using italicThreshold, not italicValue
anyway. Moved logic into editing code. For coding style reasons, added an obliqueAngle getter
and made data members private.

* Source/WebCore/css/CSSFontValue.cpp: Updated include for CSSFontStyleWithAngleValue change.
* Source/WebCore/css/CSSFontValue.h: Changed style member to CSSValue since it can now be
either a CSSPrimitiveValue or a CSSFontStyleValue.

* Source/WebCore/css/CSSProperties.json: Added all new longhands for all subproperties of
the font shorthand. Used an order that seems logical. Strangely, the order of longhands
is web-exposed so should probably be standardized and tested.

* Source/WebCore/css/CSSValue.cpp: Updated for CSSFontStyleWithAngleValue change.
* Source/WebCore/css/CSSValue.h: Ditto.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontStyle): Changed return type to CSSValue since this can return a primitive value.
Revised to only return a CSSFontStyleValue when an angle is specified.
(WebCore::fontShorthandValueForSelectionProperties): Deleted. Merged into new fontShorthandValue.
(WebCore::fontShorthandValue): Added. Implements rule for when to compute as empty, when any
longhand property is not expressible by shorthand. This separates those rules from actual
code that builds serializable value, which can now be more straightforward.
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle): Call fontShorthandValue for CSSPropertFont.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontValue const): Added a FIXME because while there are no tests
covering it, there may need to be some additional work here to make sure that font is serialized
as empty string in specified style in case where font subproperties other than the ones checked
here are set.
(WebCore::StyleProperties::getPropertyCSSValue const): Added code to return nullptr for system
font shorthands. A better solution is likely to use a different CSSValue class for these, but
that will require changes in many other places.
(WebCore::StyleProperties::asTextInternal const): Added more font subproperty longhands and
revise the comment to make it more clear that we do indeed want to support the font shorthand
once we make it so the editing code no longer relies on it.

* Source/WebCore/css/makeprop.pl: Support properties that are not enabled, but are shorthands
for a longhand that is enabled. The example of this right now is font-optical-sizing.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeSystemFont): Deleted. Merged into consumeFont.
(WebCore::CSSPropertyParser::consumeFont): Rearranged this function so all parsing comes first,
then all properties are added in a single block at the end. Added some local lambdas so that code
is shorter and much easier to read. Made this single function handle system fonts too so more code
can be shared, including all the new font subproperties.
(WebCore::CSSPropertyParser::parseShorthand): Updated to remove call to consumeSystemFont.

* Source/WebCore/css/parser/CSSPropertyParser.h: Removed consumeSystemFont.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontStyle): Changed return type to CSSValue
since this can return a primitive value. Revised to only return a CSSFontStyleWithAngleValue when
an angle is specified.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h: Updated return type of consumeFontStyle.

* Source/WebCore/editing/EditingStyle.cpp:
(WebCore::identifierForStyleProperty): Moved italic threshold logic out of CSSFontStyleValue
into this function, where it can be alongside similar bold threshold logic. Rearranged the code
a bit to be tighter, and added a comment.
(WebCore::StyleChange::extractTextStyles): Removed check for CSSValueOblique because
identifierForStyleProperty can never return CSSValueOblique. That was true before these changes
and remains true.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertFontStretchFromValue): Updated to handle system
font keyword. Turns out the correct code path was already there and only an assertion needed
to be changed.
(WebCore::Style::BuilderConverter::convertFontStyleFromValue): Updated to handle keywords
specified as CSSPrimitiveValue.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::applyFontStyle): Added. Helper so we don&apos;t copy the entire FontDescription
if we aren&apos;t changing it.
(WebCore::Style::BuilderCustom::applyInitialFontStyle): Use applyFontStyle.
(WebCore::Style::BuilderCustom::applyValueFontStyle): Ditto.
(WebCore::Style::BuilderCustom::applyValueFontStyle): Updated to handle keywords specified as
CSSPrimitiveValue, retaining support for system font shorthands. Use applyFontStyle.

* Tools/Scripts/do-webcore-rename: Used this for renaming.

Canonical link: <a href="https://commits.webkit.org/256349@main">https://commits.webkit.org/256349@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0f3b24ec1239a228e1bd00fceae4511b6f09bda7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4737 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28523 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105050 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165319 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99455 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4762 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33517 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87847 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100916 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3482 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82080 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30566 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87290 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73405 "Found 3 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20142 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4390 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40933 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42922 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39389 "Passed tests") | |

### CODE REVIEW

**Comment by darinadler** [comment]
Working on a revision to improve handling of system font

**Comment by nt1m** [comment]
@darinadler The editing failure you're seeing here may be the same as the one that was noticed when trying to add longhands to `text-decoration`. @Loirooriol had a fix that was reverted (because of a perf regression) in https://github.com/WebKit/WebKit/commit/5425a4106fd3e0acf05d21b1b2ad2c4e68f9b990#diff-95930edb6b781ea4a9c186c448fbcf9a40b69ff535c822c96c284a67fb81a89a

**Comment by darinadler** [comment]
There are many editing failures, not sure they all have exactly the same cause. But I’ll figure it out.

**Comment by darinadler** [comment]
The build failure is because `font-optical-sizing` is conditional, guarded by `ENABLE_VARIATION_FONTS`, but it’s also a subproperty of `font`. Is there some easy solution for this?

File: Source/WebCore/css/CSSProperties.json [hunk 1]
```diff
@@ -717,13 +717,24 @@
...
+                              "font-variant-alternates",
+                              "font-variant-ligatures",
+                              "font-variant-numeric",
+                              "font-variant-east-asian",
+                              "font-variant-position",
+                              "font-feature-settings",
+                              "font-optical-sizing",
+                              "font-variation-settings",
+                              "font-palette"
                               ]
```
Comment (nt1m) [inline Source/WebCore/css/CSSProperties.json]: Drive-by but can we indent this consistently while we're here?

File: Source/WebCore/css/CSSFontStyleValue.h [hunk 2]
```diff
@@ -26,45 +26,23 @@
 #pragma once
 
 #include "CSSPrimitiveValue.h"
-#include "CSSValue.h"
-#include "FontSelectionAlgorithm.h"
 
 namespace WebCore {
 
+// This class is currently only used for oblique. If we use it for more styles in the future we'll need to store the keyword.
 class CSSFontStyleValue final : public CSSValue {
```
Comment (nt1m) [inline Source/WebCore/css/CSSFontStyleValue.h]: Maybe rename this to `CSSFontStyleWithAngleValue`? It's not clear from reading this that `font-style: oblique` (with no angle) will not go through this codepath.

File: Source/WebCore/css/StyleProperties.cpp [hunk 3]
```diff
@@ -575,6 +575,8 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
 
 String StyleProperties::fontValue() const
 {
+    // FIXME: Does this need to be updated to check a lot more things now that CSSPropertyParser::consumeFont resets so much more?
+
```
Comment (nt1m) [inline Source/WebCore/css/StyleProperties.cpp]: Can we file a bug for it and reference it here? This seems to be needed, the specified style in FF & Chrome when you use a non-expressible property along with the shorthand turns into an empty string for the shorthand.

File: Source/WebCore/editing/EditingStyle.cpp [hunk 4]
```diff
@@ -148,18 +148,19 @@ static RefPtr<CSSValue> extractPropertyValue(ComputedStyleExtractor& computedSty
...
-    auto& primitiveValue = downcast<CSSPrimitiveValue>(*value);
-    if (propertyID == CSSPropertyFontWeight && primitiveValue.isNumber() && primitiveValue.doubleValue(CSSUnitType::CSS_NUMBER) >= boldThreshold())
-        return CSSValueBold;
-    return primitiveValue.valueID();
+// This synthesizes CSSValueBold and CSSValueItalic when appropriate, and never returns CSSValueOblique.
+template<typename T> CSSValueID identifierForStyleProperty(T& style, CSSPropertyID propertyID)
+{
+    auto value = extractPropertyValue(style, propertyID);
+    if (auto fontStyleValue = dynamicDowncast<CSSFontStyleValue>(value.get()))
+        return fontStyleValue->obliqueAngle().doubleValue(CSSUnitType::CSS_DEG) >= italicThreshold() ? CSSValueItalic : CSSValueNormal;
```
Comment (nt1m) [inline Source/WebCore/editing/EditingStyle.cpp]: Might be worth adding an assert that the property ID is CSSPropertyFontStyle here.

File: Source/WebCore/editing/EditingStyle.cpp [hunk 5]
```diff
@@ -148,18 +148,19 @@ static RefPtr<CSSValue> extractPropertyValue(ComputedStyleExtractor& computedSty
...
+    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value.get())) {
+        if (propertyID == CSSPropertyFontWeight && primitiveValue->doubleValue(CSSUnitType::CSS_NUMBER) >= boldThreshold())
+            return CSSValueBold;
+        auto identifier = primitiveValue->valueID();
+        return identifier == CSSValueOblique ? CSSValueItalic : identifier;
```
Comment (nt1m) [inline Source/WebCore/editing/EditingStyle.cpp]: Same here

**Review by nt1m — APPROVED** [review]
Looks great, thanks! Seems like Windows may need an extra baseline, according to EWS, since it has some unsupported properties.
